rfd/0-job-worker.md

```diff
@@ -0,0 +1,240 @@
+
+##### Identifiers
+
+JobID is an opaque identifier that is internally mapped to job in the library. Since pid is not guaranteed to be unique, nothing internal to the library will be keyed off of the pid. The library must still track the pid though in order to signal the process to stop, for example. The `go.jetify.com/typeid` library is used to generate job ids.
+
+UserID can be any unique value as far as the library is concerned. See the section on Authorization below for more information.
+
+#### gRPC API
+
+The gRPC API follows the worker library fairly closely. The only significant difference is that it does not require the user id in the messages because user id is the client certificate's subject, not explicitly provided. Additionally, `StopJob()` in the library returns a channel that closes when the job completes. The gRPC API does not provide a facility for notifying the client when the job completes.
+
+### Security Considerations
+
+#### Authentication
+
+mTLS is used for authentication. The server will be configured with a ca certificate that it expects all client certificates to be signed by. All connections presenting a certificate, signed by the configured ca certificate, are authenticated, all other connections are not.
+
+#### TLS
+
+The following TLS configuration will be used, where `clientCAs` includes the ca cert that should sign all client certs and `crt` contains the server's cert and private key. Note that cipher suites are not used because they are [not configurable](https://go.dev/blog/tls-cipher-suites) in Go when using TLS 1.3. This configuration requires and validates client certificates for all connections.
+
+```go
+tls.Config{
+    ClientAuth:   tls.RequireAndVerifyClientCert,
+    ClientCAs:    clientCAs,
+    Certificates: []tls.Certificate{crt},
+    MinVersion:   tls.VersionTLS13,
+}
+```
+
+#### Authorization
+
+The subject of the client certificate is used as a unique user identifier and is implicitly required for all requests. It is used for authorization such that only the user that starts a job is permitted to stop it, get its status or output. This way the client doesn't have to explicitly provide any other kind of identification.
+
+#### Non-Considerations
+
+At its core, this is a very dangerous service. It provides a facility for remote execution and, since it needs to be privileged to set up the namespaces, to do so as root. There are a number of things that could be done to make this safer, but are not requirements of the challenge, so they will not be implemented.
+
+Some examples of security improvements in a future version would be:
+
+- dropping privileges before executing a job
+- chroot into a separate, read-only filesystem
```

Comment:
Also hypothetical, but have you considered how safe a chroot is?

Reply:
oops, meant to write `pivot_root`. `chroot` would be bad